crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -4114,10 +4102,40 @@ impl AuthorityPerEpochStore {
         &self,
         digests: impl Iterator<Item = &'a TransactionDigest>,
     ) -> SuiResult<Vec<bool>> {
-        Ok(self
+        let size_hint = digests.size_hint().0;
+        let mut results = Vec::with_capacity(size_hint);
+        let mut fallback_keys = Vec::with_capacity(size_hint);
+        let mut fallback_indices = Vec::with_capacity(size_hint);
+
+        {
+            let consensus_quarantine = self.consensus_quarantine.read();
+
+            for (i, digest) in digests.enumerate() {
+                if consensus_quarantine.included_transaction_in_checkpoint(digest) {
+                    results.push(true);
+                } else {
+                    results.push(false);
+                    fallback_keys.push(digest);
+                    fallback_indices.push(i);
+                }
+            }
+        }
+
+        let fallback_results = self
             .tables()?
             .builder_digest_to_checkpoint
-            .multi_contains_keys(digests)?)
+            .multi_contains_keys(fallback_keys)?;
+
+        assert_eq!(fallback_results.len(), fallback_indices.len());
```

Comment:
nit: this assert is not really needed since `.zip` will panic anyway if iterators have different length

Reply:
zip does not panic! https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c92b3b8d1997c05702446e71766ae8be